### DISCUSSION TITLE
Connector Core Dump running in docker container and M2 Macbook

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/community#2] [Q&A] [answered]
Hello I am working on the FutureOn connector to iTwin.

**Problem 1:**
- I can’t get either V2 or V3 connector to run in a docker container (linux)
- Something has changed recently, because V2 used to work ok in a docker container in Kubernetes.
- My previously built version of V2 still works ok. It is just over the last couple of days where I have rebuilt the V2 and V3 versions that I have started seeing this problem.
- The code fails when downloading an iModel from iModelHub 
- It is just a single iModel that is being downloaded
- Here are some logs:
```
Info | Performance | ECDb,Schema import,Finish,1688041536998 
Info | Performance | ECDb,Profile schema import,Finish,1688041536999 
Info | Performance | ECDb,Profile schema import,Finish,1688041536999 
Info | BeSQLite | Created file '/tmp/173d9046-f3e4-4b05-8caa-af4fdb09f26a-NormalUpdate-badgers.db' 
Info | itwin-connector.Framework | Authenticating... 
Info | itwin-connector.Framework | Retrieving iModel... 
+++ loadBriefcaseDb
+++ iModelGuid 770742c6-1bba-4d66-82a5-a347b9f8bf77
+++ briefcases []
+++ bcFile undefined
Segmentation fault (core dumped)
 
The +++ are my console.logs.
```

**Problem 2:**
- I have recently upgraded to an M2 Apple Macbook
- Neither V2 or V3 unit tests run on the new macbook
- V2 gives a warning that it isn’t supported
- V3 doesn’t give any warnings.
- The V3 unit test fails on the new macbook in this code: The code fails with a **core dump**
```
  let db: SnapshotDb;
  if (!existingDbPath) {
    db = SnapshotDb.createEmpty(dbPath, {rootSubject: {name: "FutureOnBridge"}});
  }
  else {
    const seedFile = SnapshotDb.openFile(existingDbPath)
    db = SnapshotDb.createFrom(seedFile, dbPath);
  }
```

- Both V2 and V3 unit tests run on an Intel Macbook

Perhaps it is the same error that is causing the core dumps in both Problem 1 and Problem 2.

Also, is it possible to provide a meaningful error log rather than `Segmentation fault (core dumped)`?

**Comment by pmconne** [comment]
Hi @PeterJBassett.

Can you provide the text of the warning produced by V2?
In the V3 unit test snippet which branch of your conditional is being taken?
What specific version(s) of @itwin/core-backend are you using?

Segmentation faults don't generally come with helpful errors. If we can't reproduce the issue ourselves, maybe we can provide a version of the addon built with debug symbols.

**Reply by PeterJBassett** [reply]
Hello @pmconne 
Thank you for your help.
V2 gives a warning when I run `npm install`
```
npm ERR! ERROR - iModel.js cannot be installed.
npm ERR! iModel.js does not run on the arm64 version of OSX. iModel.js runs on x64 only.
```

In the V3 unit test log shown, the unit test fails in both branches, with the core dump error.
So it seems that the code fails as soon as it goes from running javascript into calling a native function. This makes me question whether the V3 library also does not support the arm64 version of OSX?

The versions of core-backend are as follows (from the package-lock.json):
V2:  bentleyjs-core-2.19.50.tgz
V3:  core-backend-3.7.9.tgz

**Reply by pmconne** [reply]
Thanks. Our [docs](https://www.itwinjs.org/learning/supportedplatforms/#backend-prerequisites) indicate arm64 is supported for macOS 11+ starting with iTwin.js 3.3.
I know we have a handful of iTwin.js developers successfully using M2 macs. I don't see evidence that we are regularly testing that architecture in our CI jobs.

What version of macOS are you running?

**Reply by PeterJBassett** [reply]
Hello @pmconne, thank you for the reply. I am using MacOS Ventura 13.4

**Reply by PeterJBassett** [reply]
Hello @pmconne - further update - I have been able to fix the Problem 2 above. This was done by deleting node_modules and package-lock.json, and setting node version to v 18 and re-installing the node_modules.

However, when installing the connector-framework, there is a warning as follows:
```
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@itwin/connector-framework@1.1.2',
npm WARN EBADENGINE   required: { node: '>=14.17.0 <17.0' },
npm WARN EBADENGINE   current: { node: 'v18.16.0', npm: '9.5.1' }
npm WARN EBADENGINE }
```

However, in the docs link you provide, it says that itwin v3 requires >= node v18. So there seems to be an incompatibility between itwin v3 and the connector-framework.

I still have had no luck fixing problem 1. But it does seem to be something with node version 16 vs 18 and compatility with the dependencies.

**Reply by PeterJBassett** [reply]
Here is some more debugging information. I ran the connector code from the command line inside the docker container, and used strace to get some information about the calls being made.
This is using the V3 connector.
```
openat(AT_FDCWD, "/app/node_modules/@bentley/imodeljs-native/imodeljs-linux-x64/Assets/ECSchemas/V3Conversion/", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 26
openat(AT_FDCWD, "/etc/localtime", O_RDONLY|O_CLOEXEC) = 26
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/blocksdb.bcv", O_RDWR|O_CREAT|O_NOFOLLOW|O_CLOEXEC, 0644) = 26
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/blocksdb.bcv-journal", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 27
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/localstore.itwindb", O_RDONLY) = 27
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/localstore.itwindb", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 27
openat(AT_FDCWD, "/openssl.cnf", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/resolv.conf", O_RDONLY) = 27
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY) = 27
openat(AT_FDCWD, "/dev/urandom", O_RDONLY) = 27
openat(AT_FDCWD, "/etc/resolv.conf", O_RDONLY) = 30
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY) = 30
openat(AT_FDCWD, "/dev/urandom", O_RDONLY) = 30
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libnss_db.so.2", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/libdb-5.3.so", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/sys/devices/system/cpu/online", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/var/lib/misc/DB_CONFIG", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/services", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/hosts", O_RDONLY) = 31
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x10} ---
--- SIGSEGV {si_signo=SIGSEGV, si_code=SI_TKILL, si_pid=67, si_uid=0} ---
+++ killed by SIGSEGV (core dumped) +++
Segmentation fault

```

**Reply by jffmarker** [reply]
We should probably focus on code based on iTwin.js 3 or 4… 2 is end-of-life. That said, 2 hasn’t been updated in over 3 months, so I’m not sure what could have changed recently. imodeljs-linux-x64 2 should be at version 2.19.50, and that’s the last I’d expect.

iTwin.js 3.x correlates to connector-framework 1.x, and iTwin.js 4.x correlates to connector-framework 2.x. I believe iTwin/FWK 3.x/1.x generally targets node 16.x, and 4.x/2.x generally targets node 18.x. There is probably some version leeway in each case, but I think most testing generally aligns to this.

Perhaps lets focus on problem 1 first... @PeterJBassett what is your Docker base image? We probably need to go down the road of trying to reproduce with a sample Connector. @pmconne, do we have any expertise collecting and symbolicating Linux core dumps to try and gather here?

CC: @akshay-madhukar @jchick-bentley

**Reply by PeterJBassett** [reply]
Hello @jffmarker here is the docker file for the V3 build.
Note that I have tried both node 16 and 18, and get the same crash.
```
FROM node:18-bullseye-slim as base
# Note - alpine-linux is not compatible with iModeljs
#        also as of 2021 the Bentley libraries require node <15

RUN mkdir /app
WORKDIR /app

## development

FROM base as development

COPY package.json package-lock.json ./
COPY logging.config.json ./
ENV NODE_PATH=/app

# updates required to solve certificate error
RUN apt-get update
RUN apt-get install -y --no-install-recommends ca-certificates git libsecret-1-dev strace libnss-db
RUN update-ca-certificates

RUN --mount=type=secret,id=npmrc,dst=/app/.npmrc \
    npm install --no-audit && npm cache clean --force

# install the generic metadata mapper (or edit to change to shell)
RUN --mount=type=secret,id=npmrc,dst=/app/.npmrc \
    npm install metadata-mapper@npm:@xvisionas/bentley-metadata-mapper-generic \
    --no-audit && npm cache clean --force


COPY . .

CMD npm run development
```

**Reply by aruniverse** [reply, marked as answer]
Can you try pinning down the version of node in your image to 18.16.0?

**Reply by PeterJBassett** [reply]
The version running in the docker container is `v18.16.1`

**Reply by PeterJBassett** [reply]
There isn't a docker base image for 18.16.0

**Reply by aruniverse** [reply]
Could you try the following?
<img width="1153" alt="image" src="https://github.com/iTwin/community/assets/11051042/b3603585-3354-48c8-a9f8-2d67e2eb4d6d">
I'm wondering if this is related to https://github.com/iTwin/itwinjs-core/issues/5663#issuecomment-1600764489

**Reply by PeterJBassett** [reply]
Hello @aruniverse have tried as you suggested node:18-bullseye and still get the same crash
Here is the `strace` 
```
openat(AT_FDCWD, "/app/node_modules/@bentley/imodeljs-native/imodeljs-linux-x64/Assets/ECSchemas/V3Conversion/", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 26
openat(AT_FDCWD, "/etc/localtime", O_RDONLY|O_CLOEXEC) = 26
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/blocksdb.bcv", O_RDWR|O_CREAT|O_NOFOLLOW|O_CLOEXEC, 0644) = 26
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/blocksdb.bcv-journal", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 27
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/localstore.itwindb", O_RDONLY) = 27
openat(AT_FDCWD, "/root/.cache/iModelJs/Workspace/cloud/localstore.itwindb", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 27
openat(AT_FDCWD, "/openssl.cnf", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/resolv.conf", O_RDONLY) = 27
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY) = 27
openat(AT_FDCWD, "/dev/urandom", O_RDONLY) = 27
openat(AT_FDCWD, "/etc/resolv.conf", O_RDONLY) = 30
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY) = 30
openat(AT_FDCWD, "/dev/urandom", O_RDONLY) = 30
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libnss_db.so.2", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/libdb-5.3.so", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/sys/devices/system/cpu/online", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/var/lib/misc/DB_CONFIG", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/services", O_RDONLY|O_CLOEXEC) = 31
openat(AT_FDCWD, "/etc/hosts", O_RDONLY) = 31
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x10} ---
--- SIGSEGV {si_signo=SIGSEGV, si_code=SI_TKILL, si_pid=71, si_uid=0} ---
+++ killed by SIGSEGV (core dumped) +++
Segmentation fault

```

**Reply by aruniverse** [reply]
i just want to confirm, you used `node:18.16.0-bullseye` and not `node:18-bullseye`

**Reply by PeterJBassett** [reply]
Hello @aruniverse - thanks you for clarifying the version to use. 
You are right, using `node:18.16.0-bullseye` rather than 'node:18-bullseye' has fixed the problem, it no longer crashes.

**Reply by aruniverse** [reply]
Thx for the confirmation; we will work on getting a patch out to fix this but in the meantime you can pin your node ver to 18.16.0
